rollouts/controllers/rollout_controller.go

```diff
@@ -294,30 +307,16 @@ func (r *RolloutReconciler) reconcileRollout(ctx context.Context, rollout *gitop
 
 	waveStatuses := []gitopsv1alpha1.WaveStatus{}
 
-	for _, wave := range strategy.Spec.Waves {
-		waveClusters, err := r.store.ListClusters(ctx, rollout.Spec.Targets.Selector, wave.Targets.Selector)
-		if err != nil {
-			return err
-		}
-
-		packageClusterMatcherClient := packageclustermatcher.NewPackageClusterMatcher(waveClusters.Items, discoveredPackages)
-		allClusterPackages, err := packageClusterMatcherClient.GetClusterPackages(rollout.Spec.PackageToTargetMatcher)
-		if err != nil {
-			logger.Error(err, "get cluster packages failed")
-			return client.IgnoreNotFound(err)
-		}
-
-		for _, clusterPackages := range allClusterPackages {
-			clusterName := clusterPackages.Cluster.Name
-			logger.Info("cluster packages", "cluster", clusterName, "packagesCount", len(clusterPackages.Packages), "packages", clusterPackages.Packages)
-		}
+	allWaveTargets, err := r.getWaveTargets(ctx, rollout, targets, strategy.Spec.Waves)
+	if err != nil {
+		return err
+	}
 
-		targets, err := r.computeTargets(ctx, rollout, allClusterPackages, waveClusters.Items)
-		if err != nil {
-			return err
-		}
+	for _, waveTargets := range allWaveTargets {
+		wave := waveTargets.Wave
+		waveTargets := waveTargets.Targets
 
-		thisWaveInProgress, clusterStatuses, err := r.rolloutTargets(ctx, rollout, &wave, targets, pauseFutureWaves)
+		thisWaveInProgress, clusterStatuses, err := r.rolloutTargets(ctx, rollout, wave, waveTargets, pauseFutureWaves)
```

Comment:
Since we are looping over and using the local variable and pointers :), we careful about the pitfall :)